File: go/candlestickpatterns/patterns/invertedhammer.go
package patterns

import (
	"zpano/candlestickpatterns/core"
	"zpano/fuzzy"
)

// InvertedHammer detects the Inverted Hammer pattern (2-candle bullish).
//
// Must have:
//   - small real body,
//   - long upper shadow,
//   - very short lower shadow,
//   - gap down from the previous candle's real body.
func InvertedHammer(cp *core.CandlestickPatterns) float64 {
	if !cp.Enough(2, cp.ShortBody, cp.LongShadow, cp.VeryShortShadow) {
		return 0.0
	}

	b1 := cp.Bar(2)
	b2 := cp.Bar(1)

	if !core.IsRealBodyGapDown(b1.O, b1.C, b2.O, b2.C) {
		return 0.0
	}

	muShort := cp.MuLess(core.RealBodyLen(b2.O, b2.C), cp.ShortBody, 1)
	muLongUS := cp.MuGreater(core.UpperShadow(b2.O, b2.H, b2.C), cp.LongShadow, 1)
	muShortLS := cp.MuLess(core.LowerShadow(b2.O, b2.L, b2.C), cp.VeryShortShadow, 1)

	confidence := fuzzy.TProductAll(muShort, muLongUS, muShortLS)
	return confidence * 100.0
}

File: ts/candlestick-patterns/patterns/inverted-hammer.ts
/** Inverted Hammer pattern (2-candle bullish). */
import { CandlestickPatternsEngine } from '../core/engine.ts';
import { tProductAll } from '../../fuzzy/index.ts';
import { realBody, upperShadow, lowerShadow, isRealBodyGapDown } from '../core/primitives.ts';

/**
 * Inverted Hammer: a two-candle bullish pattern.
 *
 * Must have:
 * - small real body,
 * - long upper shadow,
 * - very short lower shadow,
 * - gap down from the previous candle's real body.
 *
 * Returns:
 *     Continuous float in [0, 100].  Higher = stronger signal.
 */
export function invertedHammer(cp: CandlestickPatternsEngine): number {
    if (!cp.enough(2, cp.shortBody, cp.longShadow, cp.veryShortShadow)) return 0.0;

    const b1 = cp.bar(2);
    const b2 = cp.bar(1);

    if (!isRealBodyGapDown(b1.o, b1.c, b2.o, b2.c)) return 0.0;

    const muShort = cp.muLessCS(realBody(b2.o, b2.c), cp.shortBody, 1);
    const muLongUS = cp.muGreaterCS(upperShadow(b2.o, b2.h, b2.c), cp.longShadow, 1);
    const muShortLS = cp.muLessCS(lowerShadow(b2.o, b2.l, b2.c), cp.veryShortShadow, 1);

    return tProductAll(muShort, muLongUS, muShortLS) * 100.0;
}
